mantidproject/mantid
# PSIBackgroundSubtraction - Add properties StartX and EndX

**Description of work.**
This PR adds a `StartX` and `EndX` property to the PSIBackgroundSubtraction algorithm.  The background required to be subtracted will be calculated based on the X values between the `StartX` and `EndX`. The calculated background will then be subtracted across all X's.

**To test:**
Use this script for testing
```py
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

ws = CreateSampleWorkspace()
ws.setYUnit("Counts")

n_bins = ws.blocksize()
n_hist = ws.getNumberHistograms()

for i in range(n_hist):
    ws.mutableRun().addProperty("First good spectra " + str(i),  int(n_bins / 2), True)
    ws.mutableRun().addProperty("Last good spectra " + str(i),  n_bins, True)

PSIBackgroundSubtraction(InputWorkspace=ws, StartX=5000, EndX=6000)
```

1. Try StartX=5000 and EndX=6000. The resulting workspace should all be 0.3's (apart from the peak 10.3 at bin index 50). This is because the background was calculated based on there only being 0.3's between X=5000 and 6000. (meaning the background subtracted was all zero).

2. Try StartX=5000 and EndX=15000. This range encompasses the peak at bin index 50. Therefore a different background is calculated and then subtracted. The resulting workspace should all be approx -0.0058 .

3. Also if you don't provide StartX and EndX we should expect the same result as step 2. (i.e. -0.0058)



Fixes #30901

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
